walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,84 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) { 
+    // check edge cases before calling helper functions
+    if (request.getDuration() == 0 || request.getDuration() > 1440 ) { 
+        return new ArrayList<>();
+    }
+
+    List<TimeRange> unavailableTimeRanges = getUnavailableTimeRanges(events, request);
+    List<TimeRange> availableTimeRanges = getAvailableTimeRanges(unavailableTimeRanges, request.getDuration());
+
+    return availableTimeRanges;
+  }
+
+  /**
+  * Returns the time ranges for which the meeting request cannot coincide with. 
+  * A time range isn't available if there is an event happening and some people in the
+  * event are required to be in the meeting.
+  * If there are no unavailable slots, this function returns an empty arraylist.
+  *
+  * @param events the existing events with attendees that already planned
+  * @param request the meeting request containing the duration and attendees needed
+  * @return an arraylist containing TimeRanges that aren't available for the meeting, sorted by start time
+  */
+  private List<TimeRange> getUnavailableTimeRanges(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    for (Event e : events) {
+        // if there are attendees in common, that time range isn't available 
+        if (!Collections.disjoint(e.getAttendees(), request.getAttendees())) {
+            unavailableTimeRanges.add(e.getWhen());
+        }
+    }
+
+    // sort the time ranges in chronological order
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+  * Returns the time ranges for which the meeting request can happen in. 
+  * This returns the "inverse" of the unavailable timeslots.
+  * If there are no available slots, this function returns an empty arraylist.
+  *
+  * @param unavailableTimeRanges the time ranges that are not available
+  * @param duration the duration of the requested meeting
+  * @return an arraylist containing TimeRanges that are available for the meeting, sorted by start time
+  */
+  private List<TimeRange> getAvailableTimeRanges(Collection<TimeRange>unavailableTimeRanges, long duration) {
+    List<TimeRange> availableTimeRanges = new ArrayList<>();
+
+    int start = 0;
+
+    for (TimeRange tr : unavailableTimeRanges) {
+        if (tr.start() - start >= duration) {
+            availableTimeRanges.add(TimeRange.fromStartEnd(start, tr.start(), false));
+        }   
+
+        // the start time of the next available time range is the end time of the current time range
+        // or the end time of the time range in which this current time range is a subset of
+        if (tr.end() > start) {
+            start = tr.end();
+        }
+
+    }
+
+    int endOfDay = 1440;
```

Comment:
This is equivalent to `TimeRange.END_OF_DAY + 1`, or just include it as a const above.

Reply:
i'll do this instead